### PR TITLE
add method to import PEM formatted PKCS8 keys

### PR DESCRIPTION
**Pull request by Goirad**
I have added a method to import PEM formatted keys. Ideally, the other method would be changed to `import_pkcs8_der`, but that would require a major version bump. 

Also of note, windows does not validate the PEM headers, and only checks that the header matches the footer. It is unclear to me whether this crate is the place to add that logic, or whether it would be sufficient to add rudimentary checks on the input string in the `import_pkcs8_pem` method.

Another alternative would be add something like a `parse_pkcs8_pem` method and leave it to the user of this library to call that first before then calling `import_pkcs8` on the result. I would be happy to use that approach instead and include the header checking there if that is a preferred structure.

### CODE REVIEW

File: src/crypt_prov.rs [hunk 1]
```diff
@@ -283,4 +321,70 @@ mod test {
...
+            .unwrap();
+        assert!(context.import()
+            .import_pkcs8_pem(key)
+            .is_err());
+    }
+
+    #[test]
+    // FIXME this test should fail, but it does not
+    // It appears windows does not check that the header is correct,
+    // only that it matches the footer
```
Comment (jethrogb) [inline src/crypt_prov.rs]: Not sure if it's useful to have these tests here

File: src/crypt_prov.rs [hunk 2]
```diff
@@ -233,6 +234,43 @@ impl<'a> ImportOptions<'a> {
...
+    pub fn import_pkcs8_pem(&mut self, pem: &[u8]) -> io::Result<CryptKey> {
+        unsafe {
+            assert!(pem.len() <= winapi::DWORD::max_value() as usize);
+
+            // Decode the pem wrapper before passing it to import_pkcs8
+            // Call once first to figure out the necessary buffer size
+            let mut len = 0;
+            let res = wincrypt::CryptStringToBinaryA(pem.as_ptr() as ntdef::LPCSTR,
+                                                    pem.len() as winapi::DWORD,
+                                                    wincrypt::CRYPT_STRING_BASE64_ANY,
```
[the remark this answers is not in the document]
Comment (steffengy) [reply src/crypt_prov.rs]: I do not think so, parsing should work for PEM files with & without header.

File: src/crypt_prov.rs [hunk 3]
```diff
@@ -233,6 +234,43 @@ impl<'a> ImportOptions<'a> {
             res
         }
     }
+
+    /// Imports a PEM-encoded PKCS8 private key.
+    /// This functions decodes PEM blocks with or without "BEGIN PUBLIC KEY"
+    /// and "END PUBLIC KEY" headers
```
Comment (steffengy) [inline src/crypt_prov.rs]: should be "BEGIN/END PRIVATE KEY" right?

**Comment by steffengy** [comment]
I think windows only checks if the header begins with a "BEGIN".
I'd think I'd prefer the `import_pkcs8_pem` approach and to fullfill the contract of the name some rudimentary checks (if it starts with "----BEGIN" it has to be "-----BEGIN PRIVATE KEY-----" and has to end with "-----END PRIVATE KEY-----")
What do you think?

File: src/crypt_prov.rs (hunk 3 above)
Comment (Goirad) [reply src/crypt_prov.rs]: yes it should

**Comment by Goirad** [comment]
Sounds good to me, what do you think of renaming `import_pkcs8` to `import_pkcs8_der`?

**Comment by jethrogb** [comment]
What happens if you try to pass in a PKCS1 key?

**Comment by Goirad** [comment]
Turns out it rejects PKCS1 keys. I've added a test to that effect. Also it seems tests are failing for unrelated reasons.

**Comment by steffengy** [comment]
> Sounds good to me, what do you think of renaming `import_pkcs8` to `import_pkcs8_der`?

I think `import_pkcs8_pem` is better for consistency reasons. In all places I found we currently have pem explicitly as suffix and omit DER.

**Comment by Goirad** [comment]
Is there anything else that you want me to change for this PR?

**Comment by steffengy** [comment]
@Goirad It's fine from my side, maybe @sfackler wants some changes before this fits into how you intend to use it in native-tls - atleast that's what I gathered.
(Referencing https://github.com/sfackler/rust-native-tls/pull/147)

**Comment by Goirad** [comment]
@steffengy Can we get a release?

**Comment by steffengy** [comment]
@Goirad Sure, done.

**Pull request merged**